codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -319,13 +327,86 @@ private int allocateBytesForTree(Http2Stream parent, int connectionWindow) {
             totalWeight = nextTotalWeight;
-            tail = nextTail;
         }
 
         return bytesAllocated;
     }
 
     /**
+     * Utility class used by the priority algorithm to manage multiple iterations over the children of a given stream.
+     * This class will try to avoid copying the {@link Collection} returned by the parent {@link
+     * Http2Stream#children()}} by first iterating over the collection directly. As streams are found that require
+     * additional processing, they will be added to the next iteration buffer via {@link
+     * #addToNextIteration(Http2Stream)}. The next call to {@link #iterator()} will return an {@link Iterator} that will
+     * iterate only over those streams which were added back to the cache.
+     */
+    private final class ChildCache implements Iterable<Http2Stream> {
+        final Collection<? extends Http2Stream> children;
+        int nextTail;
+        Http2Stream[] childArray = null;
+
+        /**
+         * Creates the cache and initializes it to iterate over all children of the given parent in the next
+         * iteration.
+         */
+        ChildCache(Http2Stream parent) {
+            this.children = parent.children();
+            nextTail = children.size();
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public Iterator<Http2Stream> iterator() {
+            try {
+                if (nextTail == 0) {
+                    // Nothing to iterate over.
+                    return Collections.emptyIterator();
+                }
+
+                if (childArray != null) {
+                    // We've allocated the child array, iterate over it up to the specified tail position.
+                    return Iterators.newIterator(childArray, 0, nextTail);
+                }
+
+                if (nextTail != children.size()) {
+                    // We should never get here.
+                    throw new IllegalStateException(
+                            String.format("nextTail (%d) does not match num children (%d)", nextTail, children.size()));
+                }
+
+                // On the first iteration we use the children collection directly. Future iterations
+                // will use the childArray.
+                return (Iterator<Http2Stream>) children.iterator();
+            } finally {
+                // Re-initialize the next tail.
+                nextTail = 0;
+            }
+        }
+
+        /**
+         * Adds the given child to the collection iterated over by the next call to {@link #iterator()}.
+         */
+        void addToNextIteration(Http2Stream child) {
+            if (childArray == null) {
+                // Initial size is 1/4 the number of children.
+                int initialSize = min(max(4, children.size() / 4), children.size());
+                childArray = new Http2Stream[initialSize];
+            } else if (nextTail >= childArray.length) {
```

Comment:
This conditional may be a bit misleading.  If `nextTail > childArray.length` then the following resize operation may not be sufficient.  For example `nextTail = 20, childArray.length = 4, children.size() = 40`...the array would only be sized to `8` and we would have an array out of bounds....not that this can realistically happen but the point is I think we can check for `nextTail == childArray.length` because that is what this class is really designed for.